Add base client

ggreg opened the pull request:
This is the first commit that brings the initial structure of the project. The client only supports non-authenticated queries for now. The package is `prestodb` and the two main modules are:
- `client`, a low-level interface to query Presto
- `dbapi`, a DBAPI2 implementation

There are basic tests in `tests` and `integration_tests`.

electrum commented:
`CONTRIBUTING.md` is empty. Copy the one from Presto

ggreg commented:
Added license headers to all Python source files.

ggreg commented:
Added Travis config for unit tests.

electrum commented on integration_tests/fixtures.py:
```diff
@@ -0,0 +1,172 @@
+
+def download_presto_pkg(path, version=PRESTO_VERSION):
+    url = PRESTO_URL_FORMAT.format(version)
+    subprocess.check_call(['curl', '-o', path, url])
```
This will print progress. Add `-sS` to only show errors

electrum commented on integration_tests/fixtures.py:
```diff
@@ -0,0 +1,172 @@
+
+
+def untar_presto_pkg(tarball_path):
+    subprocess.check_call(['tar', 'zxf', tarball_path])
```
Technically, this should be `-zxf`

electrum commented on integration_tests/fixtures.py:
```diff
@@ -0,0 +1,172 @@
+# Beware of security issues when using files from /tmp (or other shared
+# directories) as someone else can leave there a modified version of Presto.
+PRESTO_PYTHON_CLIENT_TEST_CACHE_DIR = os.environ.get(
+    'PRESTO_PYTHON_CLIENT_TEST_CACHE_DIR',
```
Remove the trailing comma since there's always one arg:
```python
PRESTO_PYTHON_CLIENT_TEST_CACHE_DIR = os.environ.get(
    'PRESTO_PYTHON_CLIENT_TEST_CACHE_DIR')
```
Also, the local variable name is quite long. It could just be `CACHE_DIR`:
```python
CACHE_DIR = os.environ.get('PRESTO_PYTHON_CLIENT_TEST_CACHE_DIR')
```

electrum replied:
This local variable name is quite long. This could just be `CACHE_DIR`

electrum commented:
I just enabled Travis for the repo. If you force push, hopefully it will try to run the build.

electrum commented:
The license headers should be done with `#` comments, since the license is not the docstring for the module. I looked at some other Python projects and all use `#` for license header.

ggreg commented on integration_tests/fixtures.py:
```diff
@@ -0,0 +1,172 @@
+logger.setLevel(logging.INFO)
+
+PRESTO_VERSION = os.environ.get('PRESTO_PYTHON_CLIENT_PRESTO_VERSION', '0.157')
+PRESTO_URL_FORMAT = 'https://repo1.maven.org/maven2/com/facebook/presto/presto-server/{v}/presto-server-{v}.tar.gz'.format(v=PRESTO_VERSION)  # NOQA
```
Shut linters like `pylint` and `pep8`, here because this is a long line.

ggreg commented:
Thanks @electrum for the review. I am applying the changes.

ggreg commented:
I see Travis running the tests.

ggreg commented:
Passing it back to you.

ggreg commented:
I'm fixing the tests on Python2.7. Travis fails because mypy only exists for Python3. I removed it from `requirements.txt`.

electrum commented:
GitHub wraps the commit description strangely. Wrap the paragraph lines at ~70 characters to avoid this.

ggreg commented:
I am switching to *Docker* for integration tests, they run too slowing (timeout after 10 minutes) on Travis.

ggreg commented:
Have a working `Dockerfile`. Plugging the integration test on `Docker` now.

ggreg commented:
Integration test uses `Docker` now. The `Dockerfile` defines an image with Presto and a basic configuration suitable for testing. Need to update the `README`.

ggreg commented:
Travis network bandwidth looks really low to download the ~1GB Presto package. I need to find a way to workaround that.

haozhun commented:
Let me know when tests are updated and comments are addressed. This looks good mostly.

ggreg commented on prestodb/dbapi.py:
```diff
@@ -0,0 +1,231 @@
+class Connection(object):
+    """Presto does not have a notion of a persistent connection.
+
+    Thus, these objects are small stateless factories for cursors, which do all
```
I updated the comment as this is confusing and not true wrt to transactions.

ggreg commented on prestodb/dbapi.py:
```diff
@@ -0,0 +1,231 @@
+            result.append(row)
+
+        if not self._query.result.rownumber == 0:
+            raise DatabaseError('no rows returned')
```
The conversion is inverted. Agreed, that tests are lacking.

haozhun replied:
A query could legitimately return no rows. Is that handled here?

ggreg commented:
Tests work with `Docker` with *Python3.5* but they hang with *Python2.7*.

ggreg commented:
I removed Python2.7 compatibility. Most of the Python Scientific Stack (and hence most data libraries and frameworks) is focusing on [Python3](http://www.python3statement.org/).

ggreg commented:
Tests are passing and currently take 2 min to run, including integration tests.

I am adding more unit tests using Hypothesis and integration tests.

ggreg commented:
I updated the `Dockerfile` to leverage the cache. Now integration tests take around 30 seconds on my Macbook. Most of the time is spent loading the container context and starting Presto.

I added code coverage in the tests. The threshold is now 70% but I am still adding tests and will increase its value.

I also re-introduced Python2.7 compatibility as we have internal users that need it and it will also be safer for adoption by the open source community at the beginning.

ggreg commented:
I remove the coverage from Travis test runs. Python2.7 integration tests seem stalling for a long time. Trying to reproduce locally to fix them.

ggreg commented:
I added two test cases to the integration tests:
- A query that returns no rows
- A query that returns 1,000 from the TCP-H connector

I also extended `integration_tests/fixtures.py` to provide commands to manipulate Docker containers with Presto server and CLI. It is useful to experiment and do some interactive testing.

These commands are wrapped in the`run` script at the root of the project. I also added a `help` script that only reads the `README.me`.

I still plan to add more tests but for now the coverage is around 74%.

ggreg commented:
@haozhun I addressed your comments and added tests.

ggreg commented:
I'm trying to understand why the integration tests raise an error `NO_NODES_AVAILABLE` most of the time when querying TCP-H. The container runs a single process that acts as both the coordinator and the worker.

ggreg commented on prestodb/dbapi.py:
```diff
@@ -0,0 +1,232 @@
+    pass
+
+
+class DatabaseError(Error):
```
It is part of the DBAPI 2.0 specification but I just removed it as it is not used and there are other exceptions that are not defined. Error handling could be improved to better use these exceptions.

ggreg commented:
I added a function `wait_for_presto_workers()` in the `fixtures` to ensure the worker threads are active and the integration tests are deterministic.

Thanks @electrum and @haozhun for the review!
